Mathlib/Computability/Primrec.lean

```diff
@@ -1155,6 +1158,113 @@ theorem nat_strong_rec (f : α → ℕ → σ) {g : α → List σ → Option σ
+
+private theorem bindItr_primrec {l : α → List α} (hl : Primrec l) : Primrec₂ (bindItr l) :=
+  nat_rec' snd
+    (list_cons.comp fst (const []))
+    (to₂ $ list_bind (snd.comp snd) (hl.comp₂ Primrec₂.right))
+
+private theorem graph_primrec {m : α → ℕ} {l : α → List α} {g : α → List σ → Option σ}
+    (hm : Primrec m) (hl : Primrec l) (hg : Primrec₂ g) : Primrec₂ (graph m l g) :=
+  to₂ <| nat_rec' snd (const []) <|
+    to₂ <| list_bind
+      ((bindItr_primrec hl).comp
+        (fst.comp fst)
+        (nat_sub.comp (hm.comp $ fst.comp fst) (fst.comp snd))) <|
+        option_toList.comp₂ <|
+          to₂ <| option_map
+            (hg.comp snd (mapGraph_primrec.comp (snd.comp $ snd.comp fst) (hl.comp snd)))
+            (Primrec₂.pair.comp₂ (snd.comp₂ Primrec₂.left) Primrec₂.right)
+
+variable (m : α → ℕ) (f : α → σ) (l : α → List α) (g : α → List σ → Option σ)
+  (Ord : ∀ a, ∀ a' ∈ l a, m a' < m a) (H : ∀ a, g a ((l a).map f) = some (f a))
+
+private theorem bindItr_m_lt (a : α) (k) :
+    ∀ a' ∈ bindItr l a k, m a' < m a + 1 - k := by
+  induction' k with k ih <;> simp [bindItr]
+  intro a₂ a₁ ha₁ ha₂
+  have : k ≤ m a :=
+    Nat.lt_succ.mp (by simpa using Nat.add_lt_of_lt_sub $ Nat.zero_lt_of_lt (ih a₁ ha₁))
+  have : m a₁ ≤ m a - k := Nat.lt_succ.mp (by simpa [←Nat.succ_sub this] using ih a₁ ha₁)
+  exact lt_of_lt_of_le (Ord a₁ a₂ ha₂) this
+
+private theorem bindItr_eq_nil (a : α) :
+    bindItr l a (m a + 1) = [] :=
+  List.eq_nil_iff_forall_not_mem.mpr
+    (by intro a' ha'; by_contra; simpa using bindItr_m_lt m l Ord a (m a + 1) a' ha')
+
+private theorem mapGraph_graph {as as' : List α} (has : as' ⊆ as) :
+    mapGraph (as.map $ fun x => (x, f x)) as' = as'.map f := by
+  induction' as' with a as' ih <;> simp [mapGraph]
+  · have : a ∈ as ∧ as' ⊆ as := by simpa using has
+    rcases this with ⟨ha, has'⟩
+    simpa [List.lookup_graph f ha] using ih has'
+
+private theorem graph_eq_map_bindItr (a : α) (i : ℕ) (hi : i ≤ m a + 1) :
+    graph m l g a i = (bindItr l a (m a + 1 - i)).map (fun x => (x, f x)) := by
+  have graph_succ : ∀ i, graph m l g a (i + 1) =
+    (bindItr l a (m a - i)).bind fun a' =>
+      Option.toList <| (g a' $ mapGraph (graph m l g a i) (l a')).map (a', ·) := fun _ => rfl
+  have bindItr_succ : ∀ i, bindItr l a (i + 1) = (bindItr l a i).bind l := fun _ => rfl
+  induction' i with i ih
+  · symm; simpa [graph] using bindItr_eq_nil m l Ord a
+  · simp [graph_succ, bindItr_succ, ih (Nat.le_of_lt hi), Nat.succ_sub (Nat.lt_succ.mp hi)]
+    rw [List.bind_option_toList_of_eq_some]
+    intro a' ha'; simp; rw[mapGraph_graph]
+    · exact H a'
+    · exact List.subset_bind_of_mem ha' l
+
+end
+
+theorem nat_omega_rec {m : α → ℕ} (f : α → σ) {l : α → List α} {g : α → List σ → Option σ}
+    (hm : Primrec m) (hl : Primrec l) (hg : Primrec₂ g)
+    (Ord : ∀ a, ∀ a' ∈ l a, m a' < m a)
+    (H : ∀ a, g a ((l a).map f) = some (f a)) : Primrec f := by
```

Comment:
You want `f` to also depend on a parameter which plays no role. See `nat_strong_rec` which has a type `α` which is just passed to every function. (This is used in larger constructions to contain all the variables that are fixed in the recursion.)

Reply:
Proved the lemma.